applications/RomApplication/costum_strategies/rom_builder_and_solver.h

```diff
@@ -0,0 +1,811 @@
+						component_type var_component = KratosComponents< component_type >::Get(var_name);
+						auto pdof = node.pGetDof(var_component.GetSourceVariable());
+						mDofList.push_back(pdof);
+					}
+				}
+			}
+		}
+
+
+		KRATOS_CATCH("");
+	}
+
+    /**
+            organises the dofset in order to speed up the building phase
+     */
+    virtual void SetUpSystem(
+        ModelPart& r_model_part
+    ) override
+    {
+		//int free_id = 0;
+		BaseType::mEquationSystemSize = BaseType::mDofSet.size();
+		int ndofs = static_cast<int>(BaseType::mDofSet.size());
+
+#pragma omp parallel for firstprivate(ndofs)
+		for (int i = 0; i < static_cast<int>(ndofs); i++) {
+			typename DofsArrayType::iterator dof_iterator = BaseType::mDofSet.begin() + i;
+			dof_iterator->SetEquationId(i);
+
+		}
+    }
+
+    Vector ProjectToReducedBasis(const TSystemVectorType& rX, ModelPart::NodesContainerType& rNodes)
+    {
+        Vector rom_unknowns = ZeroVector(mRomDofs);
+        for(auto& node : rNodes)
+        {
+            unsigned int node_aux_id = node.GetValue(AUX_ID);
+            const Matrix& nodal_rom_basis = node.GetValue(ROM_BASIS);
+				for (int i = 0; i < mRomDofs; ++i) {
+					for (int j = 0; j < mNodalDofs; ++j) {
+						rom_unknowns[i] += nodal_rom_basis(j, i)*rX(node_aux_id*mNodalDofs + j);
+					}
+				}
+        }
+		KRATOS_WATCH(rom_unknowns)
+        return rom_unknowns;
+	}
+
+    void ProjectToFineBasis(const TSystemVectorType& rRomUnkowns, ModelPart::NodesContainerType& rNodes,  TSystemVectorType& rX)
+    {
+        TSparseSpace::SetToZero(rX);
+        for(auto& node : rNodes)
+        {
+            unsigned int node_aux_id = node.GetValue(AUX_ID);
+            const Matrix& nodal_rom_basis = node.GetValue(ROM_BASIS);
+            Vector tmp = prod(nodal_rom_basis, rRomUnkowns );
+			for (unsigned int i = 0; i < tmp.size(); ++i) 
+			{
+			rX[node_aux_id*mNodalDofs + i] = tmp[i];
+			}
+        }
+    }
+
+    void GetDofValues(const std::vector<DofPointerType>& rDofList, TSystemVectorType& rX)
+    {
+        unsigned int i=0;
+        for(auto& dof : rDofList)
+            rX[i++] = dof->GetSolutionStepValue();
+    }
+
+    /*@{ */
+
+    /**
+            Function to perform the building and solving phase at the same time.
+            It is ideally the fastest and safer function to use when it is possible to solve
+            just after building
+     */
+    virtual void BuildAndSolve(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b) override
+	{
+		
+        //define a dense matrix to hold the reduced problem
+        Matrix Arom = ZeroMatrix(mRomDofs,mRomDofs);
+        Vector brom = ZeroVector(mRomDofs);
+        TSystemVectorType x(Dx.size());
+
+        //find the rom basis
+        this->GetDofValues(mDofList,x);
+
+		double start_build4 = OpenMPUtils::GetCurrentTime();        
```

Comment:
```suggestion
        const double start_build4 = OpenMPUtils::GetCurrentTime();        
```
Why the `4`? Sounds strange to me.